### PR TITLE
fix(solutiontype): fix solution type not selected when more than one

### PR DESCRIPTION
**Pull request by stonebuzz**
If ```SELECT``` contain more than one ```options``` and if remotly sourced (AJAX)

When user choose a ```solution template```, ```solution type``` is not selected

Need to programmatically add new option with same ```id``` and ```text```

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !25222

### CODE REVIEW

File: ajax/solution.php [hunk 1]
```diff
@@ -102,5 +102,18 @@
     $template->fields['content'] = RichText::getSafeHtml($content);
 }
 
+//load solutiontype name (use to create OPTION dom)
+//need when template is used and when GLPI preselcted type if defined
```
Comment (cedric-anne) [inline ajax/solution.php:106]:
```suggestion
//need when template is used and when GLPI preselected type if defined
```

File: ajax/solution.php [hunk 2]
```diff
@@ -102,5 +102,18 @@
...
+if ($template->fields['solutiontypes_id']) {
+    $template->fields['solutiontypes_name'] = Dropdown::getDropdownName(
+        getTableForItemType(SolutionType::getType()),
+        $template->fields['solutiontypes_id'],
+        0,
+        true,
+        true,
+        "(" . $template->fields['solutiontypes_id'] . ")"
+    );//default value like "(id)" is the default behavior of GLPI when fiel 'name' is empty
```
Comment (cedric-anne) [inline ajax/solution.php:115]: Seems weird to me to have a comment located after the related code line.
```suggestion
        //default value like "(id)" is the default behavior of GLPI when field 'name' is empty
        "(" . $template->fields['solutiontypes_id'] . ")"
    );
```

**Comment by cedric-anne** [comment]
The issue only appears when the solution type used by the template is not available on the entity of the ticket. I am not sure we should permit to preselect this solution type if it is not available in the ticket entity.

**Comment by cedric-anne** [comment]
> The issue only appears when the solution type used by the template is not available on the entity of the ticket. I am not sure we should permit to preselect this solution type if it is not available in the ticket entity.

My bad, it only appears when there is more than one solution type available.

File: ajax/solution.php [hunk 3]
```diff
@@ -102,5 +102,18 @@
...
 
+//load solutiontype name (use to create OPTION dom)
+//need when template is used and when GLPI preselcted type if defined
+if ($template->fields['solutiontypes_id']) {
+    $template->fields['solutiontypes_name'] = Dropdown::getDropdownName(
+        getTableForItemType(SolutionType::getType()),
+        $template->fields['solutiontypes_id'],
+        0,
+        true,
+        true,
```
Comment (cedric-anne) [inline ajax/solution.php:113]: Not sure we need a tooltip here.
```suggestion
        false,
```

File: ajax/solution.php [hunk 4]
```diff
@@ -102,5 +102,18 @@
     $template->fields['content'] = RichText::getSafeHtml($content);
 }
 
+//load solutiontype name (use to create OPTION dom)
+//need when template is used and when GLPI preselcted type if defined
+if ($template->fields['solutiontypes_id']) {
```
Comment (cedric-anne) [inline ajax/solution.php:107]: If the solution type is not available in the item entity scope, we should unset it.

**Review by cedric-anne — CHANGES_REQUESTED** [review]
I guess same problem will also appears for the preselected category of a task template.

File: ajax/solution.php (hunk 4 above)
Comment (stonebuzz) [reply ajax/solution.php:109]: done

**Review by trasher — APPROVED** [review]

File: ajax/solution.php [hunk 5]
```diff
@@ -102,5 +102,34 @@
...
+//load solutiontype name (use to create OPTION dom)
+//need when template is used and when GLPI preselcted type if defined
+
+$template->fields['solutiontypes_name'] = "";
+if ($template->fields['solutiontypes_id']) {
+    $dbUtils = new DbUtils();
+    $entityRestrict = $dbUtils->getEntitiesRestrictCriteria(getTableForItemType(SolutionType::getType()), "", $parent->fields['entities_id'], true, false);
+    if (count($entityRestrict)) {
+        $entityRestrict = [$entityRestrict];
+    }
```
Comment (cedric-anne) [inline ajax/solution.php]:
```suggestion
    $entityRestrict = getEntitiesRestrictCriteria(getTableForItemType(SolutionType::getType()), "", $parent->fields['entities_id'], true);
```

File: more_debug.patch [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+@@ -132,6 +132,9 @@ class OperatingSystem extends InventoryAsset
+                     break;
+                 }
+             }
++            Toolbox::logError($input_os);
++            Toolbox::logError($ios->fields);
++            Toolbox::logError("Is same OS in DB -> " . $same);
+             if ($same === false) {
+                 $ios->update(['id' => $ios->getID()] + Toolbox::addslashes_deep($input_os));
+             }
```
Comment (cedric-anne) [inline more_debug.patch:14]: Should be removed.

Comment (stonebuzz) [reply more_debug.patch]: done

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Pull request merged**